### ISSUE TITLE
Different syntax for handling properties filled with interfaces or inheriting types

### ISSUE THREAD

**Issue by artyom-zuev** [aaubry/YamlDotNet#211]
Let's say we have a **Dictionary<int, **MyBaseClass**> myDictionary*\* property in a serialized class, where **MyBaseClass** is a reference type. We then have two classes inheriting from **MyBaseClass**: **MyInheritingClassA** and **MyInheritingClassB**, each with different properties, which are used for values in that dictionary. YAMLDotNet has no trouble serializing such a dictionary to this:

```
myDictionary:
      0:
        floatProperty: 2.75
        boolProperty: true
      1:
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

There is a problem with that serialization result, of course - it gives the deserializer no way of knowing that the objects in the dictionary actually aren't of **MyBaseClass** type. So the deserializer will predictably attempt to deserialize this structure into two **MyBaseClass** objects, and will stop, failing to find **floatProperty** in **MyBaseClass**. This can be solved by enabling the Roundtrip mode on the serializer, which would result in it emitting the following text:

```
myDictionary:
      0: !MyInheritingClassA,%20Assembly-CSharp,%20Version=0.0.0.0,%20Culture=neutral,%20PublicKeyToken=null
        floatProperty: 2.75
        boolProperty: true
      1: !MyInheritingClassB,%20Assembly-CSharp,%20Version=0.0.0.0,%20Culture=neutral,%20PublicKeyToken=null
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

Such an output would be perfectly deserialized into a **MyInheritingClassA** object at key 0 and **MyInheritingClassB** object at key 1. There is only one problem with this - such a verbose format isn't very friendly to manual edits and really clutters the file (if you have a collection with thousands of entries each containing some concise properties, that class info would account for 90% of text). So I started wondering if some simpler, more concise syntax can be used. Obviously, you can register custom tags on the deserializer and then easily read something like this:

```
myDictionary:
      0: !customTagA
        floatProperty: 2.75
        boolProperty: true
      1: !customTagB
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

Unfortunately, the serializer has no support for custom tags, so for something like an application config file, it's not a very useful approach - your nice tags will be removed on next serialization (or replaced with that verbose class info, as in the previous snippet if Roundtrip mode is enabled). So I started checking if it's possible to strip some parts of the Roundtrip mode type syntax. First, I tried this:

```
myDictionary:
      0: !MyInheritingClassA
        floatProperty: 2.75
        boolProperty: true
      1: !MyInheritingClassB
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

The deserializer ignored the class info written in that format and attempted to deserialize entries into **MyBaseClass**. No dice. Then I tried this:

```
myDictionary:
      0: !MyInheritingClassA,%20Assembly-CSharp
        floatProperty: 2.75
        boolProperty: true
      1: !MyInheritingClassB,%20Assembly-CSharp
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

It worked. So, apparently, version, culture and public key token parameters are optional and aren't necessary to resolve the entry to a proper class. Looks like those parameters are optional. So, I'm left wondering - why is more concise syntax not supported?

```
myDictionary:
      0: !MyInheritingClassA
        floatProperty: 2.75
        boolProperty: true
      1: !MyInheritingClassB
        integerProperty: 5
        stringProperty: "Lorem ipsum"
```

I understand that assembly, version, culture and token parameters might have some valid use cases, but it would be very nice to have an option of emitting YAML without them - just a type name over every custom type.

**Edit:** Investigated it a bit more - it looks like additional parameters aren't exactly intentional, they are just included by default into the output of the **Type.AssemblyQualifiedName** function used by the serializer in the Roundtrip mode. Replacing it with Type.Name at line 127 in TypeAssigningEventEmitter.cs turn the Rountrip-mode emitted YAML into the format I have suggested. Now to figure out how to make this deserialized.

**Edit 2:** Still not sure why the deserializer ignores the type tags containing no commas - as far as I understand, type-containing tags are handled by **TypeNameInTagNodeTypeResolver** class, and it contains nothing that would trigger than behaviour - it just grabs the **tag substring 1**, which should work with the example above. The tag might be discarded earlier, before that class gets a chance to work, I guess. Or maybe GetType called with that substring fails without any suffixes within an input, for some reason?

**Edit 3:** A-ha, looks like standard GetType returns null if the string you pass to it is not assembly-qualified. Since no info seems to be actually used from assembly parameters, GetType is probably just doing a simple check for the number of strings after a split by a comma character. Should be possible to use a different GetType method to sidestep that (maybe Assembly.GetType?). Trying that now.

**Comment by artyom-zuev** [comment]
Hell yeah, got it working. If you're interested, here are the changes required to make it work.

To get the serializer to emit type tags without assembly info:
**TypeAssigningEventEmitter.cs**, line 120:

```
        private void AssignTypeIfDifferent(ObjectEventInfo eventInfo)
        {
            if (_assignTypeWhenDifferent && eventInfo.Source.Value != null)
            {
                if (eventInfo.Source.Type != eventInfo.Source.StaticType)
                {
                    eventInfo.Tag = "!" + eventInfo.Source.Type.Name;
                    // eventInfo.Tag = "!" + eventInfo.Source.Type.AssemblyQualifiedName;
                }
            }
        }
```

To get types without assembly info recognized:
**TypeNameInTagNodeTypeResolver.cs**, line 29: 

```
bool INodeTypeResolver.Resolve(NodeEvent nodeEvent, ref Type currentType)
        {
            if (!string.IsNullOrEmpty(nodeEvent.Tag))
            {
                // If type could not be loaded, make sure to pass resolving to the next resolver
                try
                {
                    string className = nodeEvent.Tag.Substring (1);
                    System.Collections.Generic.List<Type> types = new System.Collections.Generic.List<Type> ();

                    foreach (System.Reflection.Assembly a in AppDomain.CurrentDomain.GetAssemblies ())
                    {
                        Type[] assemblyTypes = a.GetTypes ();
                        for (int j = 0; j < assemblyTypes.Length; j++)
                        {
                            if (assemblyTypes[j].Name == className)
                                types.Add (assemblyTypes[j]);
                        }
                    }

                    if (types != null && types.Count > 0)
                    {
                        currentType = types[0];
                        return true;
                    }
                }
                catch { }
            }
            return false;
        }
```

Since YAMLDotNet is a plugin, it's Assembly would be different from majority of types application code would probably want to serialize/deserialize, hence the need to grab all assemblies instead of using typeof (TypeNameInTagNodeTypeResolver) -> Assembly -> Assembly.GetType.

This can obviously be optimized to handle both assembly qualified and stripped custom type tags from the same document - e.g. make serializer check all assemblies for the same type name, and if there is more than one type, emit full assembly qualified tag instead of stripped one - and on deserializer side, use string split to see which format the tag is using, branching between old Type.GetType method and new assembly checker method appropriately). But this works as a start.

What do you think? :)

**Comment by artyom-zuev** [comment]
This would be a more optimized way of changing TypeNameInTagNodeTypeResolver.cs:

```
using System;
using System.Linq;
using System.Collections.Generic;
using YamlDotNet.Core.Events;

namespace YamlDotNet.Serialization.NodeTypeResolvers
{
    public sealed class TypeNameInTagNodeTypeResolver : INodeTypeResolver
    {
        public static List<List<Type>> typesInAssemblies;

        bool INodeTypeResolver.Resolve(NodeEvent nodeEvent, ref Type currentType)
        {
            if (!string.IsNullOrEmpty(nodeEvent.Tag))
            {
                try
                {
                    string tagContent = nodeEvent.Tag.Substring (1);

                    // If string contains a comma, then it's likely assembly qualified and we can use the simple way of retrieving the type
                    if (tagContent.Contains (','))
                    {
                        UnityEngine.Debug.Log ("YAMLDotNet | Resolving the tag as assembly qualified type name: " + tagContent);
                        currentType = Type.GetType (tagContent, true);
                        return true;
                    }

                    // If the string contains only one segment, then it's likely just a type name and we need to iterate through assemblies
                    else
                    {
                        UnityEngine.Debug.Log ("YAMLDotNet | Resolving the tag as a stripped type name: " + tagContent);
                        List<Type> typesFound = new List<Type> ();

                        if (typesInAssemblies == null)
                        {
                            typesInAssemblies = new List<List<Type>> ();
                            System.Reflection.Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies ();
                            foreach (System.Reflection.Assembly a in assemblies)
                            {
                                List<Type> types = new List<Type> (a.GetTypes ());
                                typesInAssemblies.Add (types);
                            }
                        }

                        for (int a = 0; a < typesInAssemblies.Count; ++a)
                        {
                            List<Type> types = typesInAssemblies[a];
                            for (int t = 0; t < types.Count; t++)
                            {
                                Type type = types[t];
                                if (type.Name == tagContent)
                                    typesFound.Add (type);
                            }
                        }

                        if (typesFound != null && typesFound.Count > 0)
                        {
                            if (typesFound.Count > 1)
                                UnityEngine.Debug.LogWarning ("YAMLDotNet | There are multiple types with the name " + tagContent + " over multiple assemblies: it's best to rename the class to avoid ambiguity at deserialization");

                            currentType = typesFound[0];
                            return true;
                        }
                    }
                }
                catch { }
            }
            return false;
        }
    }
}
```

No string splits, assemblies retrieved only once.

**Comment by bovender** [comment]
I am facing the same problem and wonder whether a PR would be accepted by @aaubry?

**Comment by aaubry** [comment]
I'm sorry I was unable to answer this question in a timely fashion. Pull requests are always welcome, but as you have certainly moved on to other things, I will close this issue, but feel free to reopen it if necessary.
